internal/resolution/variablesources/installed_package.go

```diff
@@ -1,76 +1,74 @@
 	catalogsort "github.com/operator-framework/operator-controller/internal/catalogmetadata/sort"
-	"github.com/operator-framework/operator-controller/internal/resolution/variables"
+	olmvariables "github.com/operator-framework/operator-controller/internal/resolution/variables"
 	"github.com/operator-framework/operator-controller/pkg/features"
 )
 
-var _ input.VariableSource = &InstalledPackageVariableSource{}
-
-type InstalledPackageVariableSource struct {
-	catalogClient BundleProvider
-	successors    successorsFunc
-	bundleImage   string
-}
-
-func (r *InstalledPackageVariableSource) GetVariables(ctx context.Context) ([]deppy.Variable, error) {
-	allBundles, err := r.catalogClient.Bundles(ctx)
-	if err != nil {
-		return nil, err
-	}
-
-	// find corresponding bundle for the installed content
-	resultSet := catalogfilter.Filter(allBundles, catalogfilter.WithBundleImage(r.bundleImage))
-	if len(resultSet) == 0 {
-		return nil, r.notFoundError()
-	}
-
-	// TODO: fast follow - we should check whether we are already supporting the channel attribute in the operator spec.
-	//       if so, we should take the value from spec of the operator CR in the owner ref of the bundle deployment.
-	//       If that channel is set, we need to update the filter above to filter by channel as well.
-	sort.SliceStable(resultSet, func(i, j int) bool {
-		return catalogsort.ByVersion(resultSet[i], resultSet[j])
-	})
-	installedBundle := resultSet[0]
-
-	upgradeEdges, err := r.successors(allBundles, installedBundle)
-	if err != nil {
-		return nil, err
-	}
-
-	// you can always upgrade to yourself, i.e. not upgrade
-	upgradeEdges = append(upgradeEdges, installedBundle)
-	return []deppy.Variable{
-		variables.NewInstalledPackageVariable(installedBundle.Package, upgradeEdges),
-	}, nil
-}
-
-func (r *InstalledPackageVariableSource) notFoundError() error {
-	return fmt.Errorf("bundleImage %q not found", r.bundleImage)
-}
-
-func NewInstalledPackageVariableSource(catalogClient BundleProvider, bundleImage string) (*InstalledPackageVariableSource, error) {
-	successors := legacySemanticsSuccessors
+// InstalledPackageVariables returns variables representing packages
+// already installed in the system.
+// Meaning that each BundleDeployment managed by operator-controller
+// has own variable.
+func InstalledPackageVariables(
+	allBundles []*catalogmetadata.Bundle,
+	bundleDeployments []rukpakv1alpha1.BundleDeployment,
+) ([]*olmvariables.InstalledPackageVariable, error) {
+	var successors successorsFunc = legacySemanticsSuccessors
 	if features.OperatorControllerFeatureGate.Enabled(features.ForceSemverUpgradeConstraints) {
 		successors = semverSuccessors
 	}
 
-	return &InstalledPackageVariableSource{
-		catalogClient: catalogClient,
-		bundleImage:   bundleImage,
-		successors:    successors,
-	}, nil
+	result := make([]*olmvariables.InstalledPackageVariable, 0, len(bundleDeployments))
+	processed := sets.Set[string]{}
```

Comment:
I'm not a big fan of using sets.Set outside of reconciler code - it creates a dependency between this package and random unrelated kube libs. Not a blocker, as this is all internal. But something to think about. Never understood why Go never bothered to provide standard impls of basic data structures but instead leaves the developer to constantly re-invent the wheel...=S

Reply:
It's [250](https://github.com/kubernetes/apimachinery/blob/master/pkg/util/sets/set.go) lines of code - if you worry about the import we could just copy it - but FWIW the package doesn't pull in anything else from Kube.

Reply:
We actively encourage the use of sets.Set anywhere/everywhere it's useful.

Reply:
> Go never bothered to provide standard impls of basic data structures but instead leaves the developer to constantly re-invent the wheel...=S

A map of empty structs is basically a set, so it kinda already exsits :) IMO map of empty structs is absolutely fine and is common in Go. But sets pacakge from k8s is ok too. I'm happy to do it eithe way.

Reply:
I think apimachinery is so ubiquitous in everything we do that it has no measurable impact to use it in a module that already has a dependency on it anyway

It also serves as a good opportunity to teach others about the fact that k8s has tons of helpful libraries scattered around in various places, with lots of usage and tests and community support/maintenance. We should be using _more_ not less, IMO.

Reply:
> A map of empty structs is basically a set, so it kinda already exsits

It is, but using maps directly requires more boilerplate (e.g. remembering to initialize a `nil` map) and the syntax can make it harder to tell that the purpose of the map is to act as a set.